documents/MIGRATION_GUIDE.md

```diff
@@ -0,0 +1,947 @@
+connectProperties.withClientId(clientId);
+builder.withConnectProperties(connectProperties);
+
+Mqtt5Client client = builder.build();
+
+// Connect to server.
+client.start();
+```
+
+### Lifecycle events
+
+Both v1 and v2 SDKs provide lifecycle events for the MQTT clients.
+
+The v1 SDK provides three lifecycle events: *on connection success*, *on connection failure*, and *on connection closed*. You can
+supply a custom callback function via subclassing `AWSIotMqttClient`. It is recommended to use lifecycle events callbacks
+to help determine the state of the MQTT client during the operation.
+
+The v2 SDK adds two new lifecycle events and provides five lifecycle events in total: *on connection success*, *on connection failure*,
+*on disconnect* (the same as *on connection closed* in the v1 SDK), *on stopped*, and *on attempting connect*. Enabling lifecycle
+events is mandatory in the v2 SDK.
+
+For more information, refer to the [MQTT5 user guide](https://github.com/aws/aws-iot-device-sdk-java-v2/blob/main/documents/MQTT5_Userguide.md#how-to-create-a-mqtt5-client).
+
+#### Example of setting lifecycle events in the v1 SDK
+
+```java
+class MyClient extends AWSIotMqttClient {
+    @Override
+    public void onConnectionSuccess() {
+    }
+
+    @Override
+    public void onConnectionFailure() {
+    }
+
+    @Override
+    public void onConnectionClosed() {
+    }
+}
+
+MyClient client = new MyClient(/*...*/);
+```
+
+#### Example of setting lifecycle events in the v2 SDK
+
+```java
+class MyLifecycleEvents implements Mqtt5ClientOptions.LifecycleEvents {
+    @Override
+    public void onAttemptingConnect(Mqtt5Client client, OnAttemptingConnectReturn onAttemptingConnectReturn) {
+    }
+
+    @Override
+    public void onConnectionSuccess(Mqtt5Client client, OnConnectionSuccessReturn onConnectionSuccessReturn) {
+    }
+    
+    @Override
+    public void onConnectionFailure(Mqtt5Client client, OnConnectionFailureReturn onConnectionFailureReturn) {
+    }
+
+    @Override
+    public void onDisconnection(Mqtt5Client client, OnDisconnectionReturn onDisconnectionReturn) {
+    }
+
+    @Override
+    public void onStopped(Mqtt5Client client, OnStoppedReturn onStoppedReturn) {
+    }
+}
+
+String clientEndpoint = "<prefix>-ats.iot.<region>.amazonaws.com";
+AwsIotMqtt5ClientBuilder builder =
+  AwsIotMqtt5ClientBuilder.newDirectMqttBuilderWithMtlsFromPath(clientEndpoint, "<certificate file path>", "<private key file path>");
+
+MyLifecycleEvents lifecycleEvents = new MyLifecycleEvents();
+builder.withLifeCycleEvents(lifecycleEvents);
+
+Mqtt5Client client = builder.build();
+```
+
+### Publish
+
+The v1 SDK provides two API calls for publishing: blocking and non-blocking. For the non-blocking version, the result of the
+publish operation is reported via a set of callbacks. If you try to publish to a topic that is not allowed by a policy,
+AWS IoT Core service will close the connection.
```

Comment:
We specifically call out what happens here (close the connection) when you try to publish to a topic that is not allowed by a policy in V1. 

We should specifically note that in the same scenario, using the v2 MQTT5 client, you do not get the connection closed but instead receive a puback with a reason code. It highlights a significant improvement because with the 5 client we get a "Not Authorized" reason code rather than a "why did I get disconnected?".

Reply:
Good catch, I added a NOTE section.